### ISSUE TITLE
Provide error code when a function call has failed

### ISSUE THREAD

**Issue by rcjsuen** [nodegit/nodegit#1206]
```JavaScript
return git.Merge.base(repo, parents[0], parents[1]);
.then(function(ancestor) {
  console.log(ancestor);
})
.catch(function(err) {
  console.log(err);
})
```
Running the code above will print out the output below if the two commits have no common ancestor.

```[Error: No merge base found]```

The error code behind this error is `Error.CODE.ENOTFOUND` but there is no way for a client to know this. Not having a common ancestor between two commits is an error that a client may be able to handle gracefully but if it the function failed because of some other fatal error then maybe the client can't do anything about it. Without an error code available in the `catch` callback, clients have no way of figuring out why the function failed outside of parsing the error message itself. It would be nice if we could have something like either of the two following patterns...

```JavaScript
catch(function(err, code) {
  console.log(err);
  console.log(code);
});
```
```JavaScript
catch(function(err) {
  console.log(err.msg);
  console.log(err.code);
});
```

**Comment by rcjsuen** [comment]
I couldn't figure out how to implement the first pattern with two arguments since the promise didn't seem to want to return two things. :(

I did get some code working that implements the second pattern though. I changed the error code's property name to be `errno` instead of `code` to match what was already used elsewhere in [`async_function.cc`](https://github.com/nodegit/nodegit/blob/38bcb340bbac946d3a389d3b13ffe1f62f0958b6/generate/templates/partials/async_function.cc#L218).

Are either of the following two proposed solutions acceptable?
If not, can anyone suggest an alternative solution that would be acceptable? Thanks!
```JavaScript
// err
{ [Error: No merge base found] errno: -3 }
// err.message
[Error: No merge base found]
// err.errno
-3
```
```JavaScript
// err
{ msg: [Error: No merge base found], errno: -3 }
// err.msg
[Error: No merge base found]
// err.errno
-3
```

**Comment by maxkorp** [comment]
I like the second proposal for sure.
CC: @johnhaley81 @tbranyen Any thoughts?

**Comment by tbranyen** [comment]
In order to pass two values to a `then` you'd have to resolve an array and then use destructuring.

``` js
catch(function([ err, code ]) {
  console.log(err);
  console.log(code);
});
```

That said, I much prefer the second approach, except why is there `msg`? Why can't you just attach the `errno` directly to the error object?

**Comment by rcjsuen** [comment]
> Why can't you just attach the errno directly to the error object?

We can do that, @tbranyen. That is what is shown in the first example of my [second comment](https://github.com/nodegit/nodegit/issues/1206#issuecomment-276053704).

**Comment by rcjsuen** [comment]
I have pushed PR #1209 to expose the error code to NodeGit clients.

**Comment by tbranyen** [comment]
Just wanted to leave this internal Node error code reference: https://github.com/nodejs/node/blob/master/doc/guides/using-internal-errors.md

**Comment by rcjsuen** [comment]
Closing as PR #1209 has been merged in. Will be included in 0.18.0.
